contracts/Registry.vy

```diff
@@ -0,0 +1,253 @@
+    rewards: address,
+    guardian: address,
+    name: String[64],
+    symbol: String[32],
+) -> address:
+
+    # NOTE: Underflow if no releases created yet (this is okay)
+    vault: address = create_forwarder_to(template)
+
+    nameOverride: String[64] = name
+    if nameOverride == "":
+        nameOverride = concat("Yearn ", DetailedERC20(token).name(), " Vault")
+
+    symbolOverride: String[32] = symbol
+    if symbolOverride == "":
+        symbolOverride = concat("yv", DetailedERC20(token).symbol())
+
+    # NOTE: Must initialize the Vault atomically with deploying it
+    Vault(vault).initialize(token, governance, rewards, nameOverride, symbolOverride, guardian)
+
+    return vault
+
+
+@external
+def newVault(
+    token: address,
+    guardian: address,
+    rewards: address,
+    name: String[64],
+    symbol: String[32],
+) -> address:
+    """
+    @notice
+        Add a new deployed vault for the given token as a simple "forwarder-style" proxy to the
+        latest version being managed by this registry
+    @dev
+        If `nameOverride` is not specified, the name will be 'yearn' combined with the name of
+        `token`.
+
+        If `symbolOverride` is not specified, the symbol will be 'yv' combined with the symbol
+        of `token`.
+    @param token The token that may be deposited into this Vault.
+    @param guardian The address authorized for guardian interactions.
+    @param rewards The address to use for collecting rewards.
+    @param name Specify a custom Vault name. Set to empty string for default choice.
+    @param symbol Specify a custom Vault symbol name. Set to empty string for default choice.
+    @return The address of the newly-deployed vault
+    """
+    assert msg.sender == self.governance  # dev: unauthorized
+
+    # NOTE: Underflow if no releases created yet (this is okay)
+    release_template: address = self.releases[self.nextRelease - 1]  # dev: no releases
+    vault: address = self._newVault(token, release_template, msg.sender, msg.sender, guardian, name, symbol)
+
+    self._endorseVault(token, vault)
+
+    return vault
+
+
+@external
+def newExperimentalVault(
+    token: address,
+    governance: address,
+    guardian: address,
+    rewards: address,
+    name: String[64],
+    symbol: String[32],
+) -> address:
+    # NOTE: Anyone can call this method, as a convenience to Strategist' experiments
+    # NOTE: Underflow if no releases created yet (this is okay)
+    release_template: address = self.releases[self.nextRelease - 1]  # dev: no releases
+    vault: address = self._newVault(token, release_template, governance, governance, guardian, name, symbol)
+    # NOTE: Must initialize the Vault atomically with deploying it
+    Vault(vault).initialize(token, governance, rewards, name, symbol, guardian)
+
+    # NOTE: Don't add to list of endorsed vaults (hence no event there, so we emit here)
+    log NewExperimentalVault(token, vault, Vault(release_template).apiVersion())
```

Comment:
do we want to index also the msg.sender, creator of the vault?

Reply:
solved in 6216fda